flaking e2e tests on RUN go mod download

<!--
Please use this template for reporting flaky tests.
Links to specific failures in Prow are appreciated.
-->

**Which test is flaking?**:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/12016/pull-kueue-test-e2e-multikueue-baseline-main/2064285838954467328

**Link to failed CI job or steps to reproduce locally**:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/12016/pull-kueue-test-e2e-multikueue-baseline-main/2064285838954467328

**Failure message or logs**:
```

#8 [builder 2/6] WORKDIR /workspace
#8 DONE 0.2s
#9 [builder 3/6] COPY go.mod go.sum ./
#9 DONE 0.0s
#10 [builder 4/6] RUN go mod download
#10 5.897 go: k8s.io/apiserver@v0.36.1: read "https://proxy.golang.org/k8s.io/apiserver/@v/v0.36.1.zip": http2: server sent GOAWAY and closed the connection; LastStreamID=1013, ErrCode=NO_ERROR, debug="server_shutting_down"
#10 5.897 go: k8s.io/autoscaler/cluster-autoscaler/apis@v0.0.0-20240830133931-adfbe7e45277: Get "https://proxy.golang.org/k8s.io/autoscaler/cluster-autoscaler/apis/@v/v0.0.0-20240830133931-adfbe7e45277.zip": http2: server sent GOAWAY and closed the connection; LastStreamID=1013, ErrCode=NO_ERROR, debug="server_shutting_down"
#10 ERROR: process "/bin/sh -c go mod download" did not complete successfully: exit code: 1
------
 > [builder 4/6] RUN go mod download:
5.897 go: k8s.io/apiserver@v0.36.1: read "https://proxy.golang.org/k8s.io/apiserver/@v/v0.36.1.zip": http2: server sent GOAWAY and closed the connection; LastStreamID=1013, ErrCode=NO_ERROR, debug="server_shutting_down"
5.897 go: k8s.io/autoscaler/cluster-autoscaler/apis@v0.0.0-20240830133931-adfbe7e45277: Get "https://proxy.golang.org/k8s.io/autoscaler/cluster-autoscaler/apis/@v/v0.0.0-20240830133931-adfbe7e45277.zip": http2: server sent GOAWAY and closed the connection; LastStreamID=1013, ErrCode=NO_ERROR, debug="server_shutting_down"
------
Dockerfile:9
--------------------
   7 |     # fetch dependencies first, for iterative development
   8 |     COPY go.mod go.sum ./
   9 | >>> RUN go mod download
  10 |     # copy the rest of the sources and build
  11 |     COPY . .
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c go mod download" did not complete successfully: exit code: 1
make: *** [Makefile:232: image-build] Error 1
+ EXIT_VALUE=2
+ set +o xtrace
```

**Anything else we need to know?**:

Looks like another case which could have been prevented: https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/periodic-kueue-test-e2e-tas-baseline-release-0-18/2064550560266194944

here we have:

```
cd /home/prow/go/src/kubernetes-sigs/kueue/hack/tools && go mod download && \
 go build -ldflags="-X 'sigs.k8s.io/kueue/pkg/version.GitVersion=v0.18.0-36-g46034da4c' -X 'sigs.k8s.io/kueue/pkg/version.GitCommit=46034da4c23d63b3a6880108a8d8ef1a12e189dd' -X 'sigs.k8s.io/kueue/pkg/version.BuildDate=2026-06-10T03:34:34Z'" -o /home/prow/go/src/kubernetes-sigs/kueue/bin/ginkgo-top ./ginkgo-top
go: github.com/MirrexOne/unqueryvet@v1.5.4: reading https://proxy.golang.org/github.com/%21mirrex%21one/unqueryvet/@v/v1.5.4.zip: 500 Internal Server Error
make: *** [Makefile-test.mk:681: ginkgo-top] Error 1
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -6,7 +6,8 @@ FROM --platform=${BUILDPLATFORM} ${BUILDER_IMAGE} AS builder
 WORKDIR /workspace
 # fetch dependencies first, for iterative development
 COPY go.mod go.sum ./
-RUN go mod download
+COPY hack/testing/retry.sh /usr/local/bin/retry.sh
+RUN retry.sh --attempts 7 --delay 2 --exponential --stream -- go mod download
 # copy the rest of the sources and build
 COPY . .
 ARG GIT_TAG GIT_COMMIT TARGETARCH CGO_ENABLED=0

--- a/Makefile
+++ b/Makefile
@@ -184,12 +184,12 @@ fmt: ## Run go fmt against code.
 .PHONY: gomod-download
 gomod-download: ## Download Go module dependencies (main)
 	@echo "→ Downloading main dependencies..."
-	$(GO_CMD) mod download
+	$(NETWORK_INSTALL_RETRY) $(GO_CMD) mod download
 
 .PHONY: gomod-download-tools
 gomod-download-tools: ## Download Go module dependencies (tools)
 	@echo "→ Downloading tools dependencies..."
-	cd $(TOOLS_DIR) && $(GO_CMD) mod download
+	cd $(TOOLS_DIR) && $(NETWORK_INSTALL_RETRY) $(GO_CMD) mod download
 
 .PHONY: toc-update
 toc-update: mdtoc
@@ -440,7 +440,7 @@ kueueviz-image-build:
 		--build-arg CGO_ENABLED=$(CGO_ENABLED) \
 		$(PUSH) \
 		$(IMAGE_BUILD_EXTRA_OPTS) \
-		-f ./cmd/kueueviz/backend/Dockerfile ./cmd/kueueviz/backend
+		-f ./cmd/kueueviz/backend/Dockerfile .
 	$(IMAGE_BUILD_CMD) \
 		-t $(IMAGE_TAG_KUEUEVIZ_FRONTEND) \
 		-t $(IMAGE_REPO_KUEUEVIZ_FRONTEND):$(RELEASE_BRANCH) \

--- a/Makefile-test.mk
+++ b/Makefile-test.mk
@@ -728,7 +728,7 @@ run-large-scale-performance-scheduler-in-cluster: envtest performance-scheduler-
 
 .PHONY: ginkgo-top
 ginkgo-top:
-	cd $(TOOLS_DIR) && go mod download && \
+	cd $(TOOLS_DIR) && $(NETWORK_INSTALL_RETRY) $(GO_CMD) mod download && \
 	$(GO_BUILD_ENV) $(GO_CMD) build -ldflags="$(LD_FLAGS)" -o $(BIN_DIR)/ginkgo-top ./ginkgo-top
 
 .PHONY: setup-e2e-env

--- a/cmd/experimental/kueue-populator/Dockerfile
+++ b/cmd/experimental/kueue-populator/Dockerfile
@@ -16,7 +16,8 @@ COPY . .
 
 WORKDIR /workspace/cmd/experimental/kueue-populator
 
-RUN go mod download
+COPY hack/testing/retry.sh /usr/local/bin/retry.sh
+RUN retry.sh --attempts 7 --delay 2 --exponential --stream -- go mod download
 RUN make build GO_BUILD_ENV='CGO_ENABLED=${CGO_ENABLED} GOOS=linux GOARCH=${TARGETARCH}'
 
 FROM ${BASE_IMAGE}

--- a/cmd/experimental/kueue-priority-booster/Dockerfile
+++ b/cmd/experimental/kueue-priority-booster/Dockerfile
@@ -13,7 +13,8 @@ COPY . .
 
 WORKDIR /workspace/cmd/experimental/kueue-priority-booster
 
-RUN go mod download
+COPY hack/testing/retry.sh /usr/local/bin/retry.sh
+RUN retry.sh --attempts 7 --delay 2 --exponential --stream -- go mod download
 RUN make build GO_BUILD_ENV='CGO_ENABLED=${CGO_ENABLED} GOOS=linux GOARCH=${TARGETARCH}'
 
 FROM ${BASE_IMAGE}

--- a/cmd/kueueviz/backend/Dockerfile
+++ b/cmd/kueueviz/backend/Dockerfile
@@ -3,12 +3,12 @@ ARG BUILDER_IMAGE=golang:1.26
 ARG BASE_IMAGE=gcr.io/distroless/static:nonroot
 FROM --platform=${BUILDPLATFORM} ${BUILDER_IMAGE} AS builder
 
-# Copy Go modules manifests and download dependencies
-COPY go.mod go.sum ./
-RUN go mod download
+COPY cmd/kueueviz/backend/go.mod cmd/kueueviz/backend/go.sum ./
+COPY hack/testing/retry.sh /usr/local/bin/retry.sh
+RUN retry.sh --attempts 7 --delay 2 --exponential --stream -- go mod download
 
 # Copy the application source code
-COPY . .
+COPY cmd/kueueviz/backend/ .
 
 # Build the application
 ARG TARGETARCH
@@ -28,5 +28,3 @@ EXPOSE 8080
 
 # Command to run the application
 ENTRYPOINT ["/kueueviz"]
-
-

